frontend/catalyst/jax_tracer.py

```diff
@@ -528,11 +528,27 @@ def trace_quantum_function(
         # (1) - Classical tracing
         quantum_tape = QuantumTape()
         with EvaluationContext.frame_tracing_context(ctx) as trace:
-            wffa, in_avals, out_tree_promise = deduce_avals(f, args, kwargs)
+            wffa, in_avals, _ = deduce_avals(f, args, kwargs)
             in_classical_tracers = _input_type_to_tracers(trace.new_arg, in_avals)
             with QueuingManager.stop_recording(), quantum_tape:
                 # Quantum tape transformations happen at the end of tracing
                 ans = wffa.call_wrapped(*in_classical_tracers)
+
+            # Ans contains the leaves of the pytree (empty for measurement without
+            # data https://github.com/PennyLaneAI/pennylane/pull/4607)
+            # Therefore we need to compute the tree with measurements as leaves and it comes
+            # with an extra computational cost
+
+            # 1. Recompute the original return
+            with QueuingManager.stop_recording():
+                ans = tree_unflatten(wffa.stores[0].val, ans)
+
+            def is_leaf(obj):
+                return isinstance(obj, qml.measurements.MeasurementProcess)
+
+            # 2. Create a new tree that has measurements as leaves
+            ans, out_tree = jax.tree_util.tree_flatten(ans, is_leaf=is_leaf)
```

Comment:
Does this function create issues when non-measurements are returned from the function? For example, if a jax array is returned, which should be a leaf, this function will say it is not. Would that raise an error?

Reply:
No it is more like a custom leaf but it does not impact non measurement process.